pkg/querier/querier.go

```diff
@@ -193,53 +194,92 @@ func (q *Querier) LabelNames(ctx context.Context, req *connect.Request[typesv1.L
 	}), nil
 }
 
-func filterSeriesByLabelNames(labelSets []*typesv1.Labels, labelNameMap map[string]struct{}) {
-	if len(labelNameMap) == 0 {
-		return
-	}
-
-	for _, ls := range labelSets {
-		labelCount := 0
-		for idx := range ls.Labels {
-			_, ok := labelNameMap[ls.Labels[idx].Name]
-			if ok {
-				ls.Labels[labelCount] = ls.Labels[idx]
-				labelCount++
-			}
-		}
-		ls.Labels = ls.Labels[:labelCount]
-	}
-}
-
 func (q *Querier) Series(ctx context.Context, req *connect.Request[querierv1.SeriesRequest]) (*connect.Response[querierv1.SeriesResponse], error) {
 	sp, ctx := opentracing.StartSpanFromContext(ctx, "Series")
-	defer func() {
-		sp.LogFields(
-			otlog.String("matchers", strings.Join(req.Msg.Matchers, ",")),
-			otlog.String("label_names", strings.Join(req.Msg.LabelNames, ",")),
-		)
-		sp.Finish()
-	}()
+	defer sp.Finish()
 
-	// build up map of label names
-	labelNameMap := make(map[string]struct{}, len(req.Msg.LabelNames))
-	for _, labelName := range req.Msg.LabelNames {
-		labelNameMap[labelName] = struct{}{}
-	}
+	sp.LogFields(
+		otlog.String("matchers", strings.Join(req.Msg.Matchers, ",")),
+		otlog.String("label_names", strings.Join(req.Msg.LabelNames, ",")),
+		otlog.Int64("start", req.Msg.Start),
+		otlog.Int64("end", req.Msg.End),
+	)
 
-	responses, err := forAllIngesters(ctx, q.ingesterQuerier, func(childCtx context.Context, ic IngesterQueryClient) ([]*typesv1.Labels, error) {
-		res, err := ic.Series(childCtx, connect.NewRequest(&ingestv1.SeriesRequest{
+	// Some clients may not be sending us timestamps. If start or end are 0, then
+	// mark this a legacy request. Legacy requests only query the ingesters.
+	legacyRequest := req.Msg.Start == 0 || req.Msg.End == 0
+	sp.LogFields(otlog.Bool("legacy_request", legacyRequest))
+	if q.storeGatewayQuerier == nil || legacyRequest {
+		responses, err := q.seriesFromIngesters(ctx, &ingestv1.SeriesRequest{
 			Matchers:   req.Msg.Matchers,
 			LabelNames: req.Msg.LabelNames,
-		}))
+			Start:      req.Msg.Start,
+			End:        req.Msg.End,
+		})
 		if err != nil {
 			return nil, err
 		}
-		// TODO: Remove this once we have the ingesters returning the filtered response, has been rolled out (f25).
-		filterSeriesByLabelNames(res.Msg.LabelsSet, labelNameMap)
 
-		return res.Msg.LabelsSet, nil
-	})
+		return connect.NewResponse(&querierv1.SeriesResponse{
+			LabelsSet: lo.UniqBy(
```

Comment:
UniqBy does not sort the slice (as far I can tell). The Head querier returned sorted labels, I think that would be better for consistency.

Reply:
Fixed in https://github.com/grafana/pyroscope/pull/2431/commits/59f4ed20bfd55a39ac4df2204e3708fe751c9355